### PR TITLE
build: add additional code guideline checks

### LINKED ISSUE #58: build: add additional code guideline checks
To ensure our codebase is well structured and adheres to common python styles.

### PR DESCRIPTION
**Pull request by sasanjac**
Fixes #58 #79

### CODE REVIEW

**Comment by sasanjac** [comment]
Don't merge this until the discussion in #79 is completed.

File: powerfactory_tools/schema/steadystate_case/controller.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    class QmaxOENotSpecifiedError(ValueError):
+        def __init__(self) -> None:
+            super().__init__("qmax_oe must be specified for Q(U)-characteristic-control.")
+
+    class QmaxUENotSpecifiedError(ValueError):
+        def __init__(self) -> None:
+            super().__init__("qmax_ue must be specified for Q(U)-characteristic-control.")
+
+    class TabTarNotSpecifiedError(ValueError):
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/controller.py:70]:
```suggestion
    class QUSlopeNotSpecifiedError(ValueError):
```

File: powerfactory_tools/schema/steadystate_case/controller.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+class Controller(Base):
+    controller_type: ControllerType
+    external_controller_name: str | None = None  # if external controller is specified --> name
+    # cos(phi) control mode
+    cosphi_type: CosphiDir | None = None  # CosphiDir
+    cosphi: float | None = None  # cos(phi) for calculation of Q in relation to P.
+    # q-setpoint control mode
+    q_set: float | None = None  # Setpoint of reactive power.
+    # Q(U) characteristic control mode
+    m_tab2015: float | None = None  # Droop based on VDE-AR-N 4120:2015: '%'-value --> Q = m_% * Pr * dU_kV
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/controller.py:84]:
```suggestion
    m_tab2015: float | None = None  # Droop/Slope based on VDE-AR-N 4120:2015: '%'-value --> Q = m_% * Pr * dU_kV
```

File: powerfactory_tools/schema/steadystate_case/controller.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    m_tar2018: float | None = None  # Droop based on VDE-AR-N 4120:2018: '%'-value --> Q = m_% * Pr * dU_(% of Un)
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/controller.py:85]:
```suggestion
    m_tar2018: float | None = None  # Droop/Slope based on VDE-AR-N 4120:2018: '%'-value --> Q = m_% * Pr * dU_(% of Un)
```

File: powerfactory_tools/schema/steadystate_case/controller.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        raise Exceptions.UdeadbandLowNotSpecifiedError
+
+    if values["qmax_oe"] is None:
+        raise Exceptions.QmaxOENotSpecifiedError
+
+    if values["qmax_ue"] is None:
+        raise Exceptions.QmaxUENotSpecifiedError
+
+    if values["m_tab2015"] is None and values["m_tar2018"] is None:
+        raise Exceptions.TabTarNotSpecifiedError
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/controller.py:190]:
```suggestion
        raise Exceptions.QUSlopeNotSpecifiedError
```

File: powerfactory_tools/schema/steadystate_case/controller.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    cosphi: float | None = None  # cos(phi) for calculation of Q in relation to P.
+    # q-setpoint control mode
+    q_set: float | None = None  # Setpoint of reactive power.
+    # Q(U) characteristic control mode
+    m_tab2015: float | None = None  # Droop based on VDE-AR-N 4120:2015: '%'-value --> Q = m_% * Pr * dU_kV
+    m_tar2018: float | None = None  # Droop based on VDE-AR-N 4120:2018: '%'-value --> Q = m_% * Pr * dU_(% of Un)
+    u_q0: float | None = None  # Voltage value, where Q=0: per unit value related to Un
+    udeadband_up: float | None = None  # Width of upper deadband (U_1_up - U_Q0): per unit value related to Un
+    udeadband_low: float | None = None  # Width of lower deadband (U_Q0 - U_1_low): per unit value related to Un
+    qmax_ue: float | None = None  # Over excited limit of Q: absolut per unit value related to Pr
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/controller.py:89]:
```suggestion
    qmax_ue: float | None = None  # Over excited limit of Q: absolut value
```

File: powerfactory_tools/exporter/exporter.py [hunk 6]
```diff
@@ -0,0 +1,2320 @@
...
+            "cpchar": ControllerType.COSPHI_P,
+        }
+        s_r = gen.sgn
+        cosphi_r = gen.cosn
+        q_r = s_r * math.sin(math.acos(cosphi_r))
+
+        cosphi_type = None
+        cosphi = None
+        q_set = None
+        m_tab2015 = None  # Q(U) droop related to VDE-AR-N 4120:2015
```
Comment (SebastianDD) [inline powerfactory_tools/exporter/exporter.py:1904]:
```suggestion
        m_tab2015 = None  # Q(U) droop/slope related to VDE-AR-N 4120:2015
```

File: powerfactory_tools/exporter/exporter.py [hunk 7]
```diff
@@ -0,0 +1,2320 @@
...
+        m_tar2018 = None  # Q(U) droop related to VDE-AR-N 4120:2018
```
Comment (SebastianDD) [inline powerfactory_tools/exporter/exporter.py:1905]:
```suggestion
        m_tar2018 = None  # Q(U) droop/slope related to VDE-AR-N 4120:2018
```

File: powerfactory_tools/exporter/exporter.py [hunk 8]
```diff
@@ -0,0 +1,2320 @@
...
+        if ext_ctrl is None:
+            ext_ctrl_name = None
+            controller_type = controller_type_dict_default[gen.av_mode]
+            if controller_type == ControllerType.COSPHI_CONST:
+                cosphi = gen.cosgini
+                cosphi_type = CosphiDir.UE if gen.pf_recap == 1 else CosphiDir.OE
+            elif controller_type == ControllerType.Q_CONST:
+                q_set = gen.qgini
+            elif controller_type == ControllerType.Q_U:
+                qmax_ue = abs(gen.Qfu_min)
```
Comment (SebastianDD) [inline powerfactory_tools/exporter/exporter.py:1921]:
```suggestion
                qmax_ue = abs(gen.Qfu_min)  # absolute value
```

File: powerfactory_tools/exporter/exporter.py [hunk 9]
```diff
@@ -0,0 +1,2320 @@
...
+                qmax_oe = abs(gen.Qfu_max)
```
Comment (SebastianDD) [inline powerfactory_tools/exporter/exporter.py:1922]:
```suggestion
                qmax_oe = abs(gen.Qfu_max)  # absolute value
```

File: powerfactory_tools/schema/topology/external_grid.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+from __future__ import annotations
+
+from enum import Enum
+
+from powerfactory_tools.schema.base import Base
+
+
+class GridType(Enum):
+    SL = "SL"
```
Comment (SebastianDD) [inline powerfactory_tools/schema/topology/external_grid.py:14]:
```suggestion
    SL = "SL"  # slack node: voltage amplitude and phase angle is fixed
```

File: powerfactory_tools/schema/topology/external_grid.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    PV = "PV"
```
Comment (SebastianDD) [inline powerfactory_tools/schema/topology/external_grid.py:15]:
```suggestion
    PV = "PV"  # active power and voltage amplitude is fixed
```

File: powerfactory_tools/schema/topology/external_grid.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    PQ = "PQ"
```
Comment (SebastianDD) [inline powerfactory_tools/schema/topology/external_grid.py:16]:
```suggestion
    PQ = "PQ"  # active power and reactive power is fixed
```

File: powerfactory_tools/schema/steadystate_case/case.py [hunk 13]
```diff
@@ -99,7 +99,8 @@ def _is_proper_external_grids(self, topology: Topology) -> bool:
         for ext_grid in topology.external_grids:
             if ext_grid.name not in external_grid_names:
                 logger.error(
-                    "External grid {ext_grid_name} is not in steadystate case.", extra={"ext_grid_name": ext_grid.name},
+                    "External grid {ext_grid_name} is not in steadystate case.",
+                    extra={"ext_grid_name": ext_grid.name},
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/case.py:102]: , ext_grid_name=ext_grid.name

**Comment by SebastianDD** [comment]
within `create_topology()`

Shouldn't be return all called functions (`create_external_grids`, `create_nodes`, `create_branches`, `create_loads`, `create_transformers`) set objects instead of a mixture of sets and lists?

e.g. within create_loads() all different load types are computed as set and are then converted in a list to be returned

same for ssc. I'm not sure if this is best practice.

**Comment by SebastianDD** [comment]
still unhashable type: 'dict' when tries to write topology to json

**Comment by sasanjac** [comment]
Yeah thas a bug in `pydantic`, see https://github.com/pydantic/pydantic/issues/1090

**Review by SebastianDD — APPROVED** [review]

**Pull request merged**